File: zeitgeist-protocol/ts/test_zeitgeist.ts
/**
 * Tests for Zeitgeist Protocol — TypeScript implementation
 * Run with: npx tsx test_zeitgeist.ts
 */

import {
  Zeitgeist, PrecisionState, ConfidenceState, TrajectoryState,
  ConsensusState, TemporalState, Trend, Phase,
} from "./zeitgeist.ts";

function randRange(min: number, max: number): number {
  return Math.random() * (max - min) + min;
}

function randomZeitgeist(): Zeitgeist {
  const bloom = new Uint8Array(32);
  for (let i = 0; i < 32; i++) bloom[i] = Math.floor(Math.random() * 256);

  const crdt = new Map<number, number>();
  for (let i = 0; i < Math.floor(Math.random() * 5); i++) {
    crdt.set(Math.floor(Math.random() * 1e18), Math.floor(Math.random() * 1e18));
  }

  return new Zeitgeist(
    new PrecisionState(randRange(0.001, 99999), randRange(0, 1), Math.random() > 0.5),
    new ConfidenceState(bloom, Math.floor(Math.random() * 256), randRange(0, 1)),
    new TrajectoryState(
      randRange(0, 1),
      [Trend.STABLE, Trend.RISING, Trend.FALLING, Trend.CHAOTIC][Math.floor(Math.random() * 4)],
      randRange(-10, 10),
    ),
    new ConsensusState(randRange(0, 1), randRange(0, 1), crdt),
    new TemporalState(
      randRange(0, 1),
      [Phase.IDLE, Phase.APPROACHING, Phase.SNAP, Phase.HOLD][Math.floor(Math.random() * 4)],
      randRange(0, 1),
    ),
  );
}

// ── Merge law tests ────────────────────────────────────────

let passed = 0;
let failed = 0;

function assert(condition: boolean, msg: string) {
  if (!condition) {
    console.error(`FAIL: ${msg}`);
    failed++;
  } else {
    passed++;
  }
}

// Commutativity: 100 iterations
for (let i = 0; i < 100; i++) {
  const a = randomZeitgeist();
  const b = randomZeitgeist();
  assert(a.merge(b).equals(b.merge(a)), `Commutativity violated at iteration ${i}`);
}

// Associativity: 100 iterations
for (let i = 0; i < 100; i++) {
  const a = randomZeitgeist();
  const b = randomZeitgeist();
  const c = randomZeitgeist();
  assert(
    a.merge(b).merge(c).equals(a.merge(b.merge(c))),
    `Associativity violated at iteration ${i}`,
  );
}

// Idempotency: 100 iterations
for (let i = 0; i < 100; i++) {
  const a = randomZeitgeist();
  assert(a.merge(a).equals(a), `Idempotency violated at iteration ${i}`);
}

// JSON roundtrip: 50 iterations
for (let i = 0; i < 50; i++) {
  const zg = randomZeitgeist();
  const encoded = zg.encode();
  const decoded = Zeitgeist.decode(encoded);
  assert(decoded.equals(zg), `JSON roundtrip failed at iteration ${i}`);
}

// Alignment valid
{
  const zg = randomZeitgeist();
  const report = zg.checkAlignment();
  assert(report.aligned, `Random ZG should be aligned: ${report.violations.join(", ")}`);
}

// Alignment detects violations
{
  const zg = randomZeitgeist();
  zg.precision.deadband = -1;
  zg.confidence.certainty = 2;
  zg.trajectory.hurst = 1.5;
  zg.temporal.beat_pos = -0.5;
  const report = zg.checkAlignment();
  assert(!report.aligned, "Should detect violations");
  assert(report.violations.length >= 4, `Expected 4+ violations, got ${report.violations.length}`);
}

console.log(`\nResults: ${passed} passed, ${failed} failed`);
if (failed > 0) process.exit(1);

File: zeitgeist-protocol/python/test_zeitgeist.py
"""Tests for Zeitgeist Protocol — Python implementation"""

import random
import unittest
from zeitgeist import (
    Zeitgeist, PrecisionState, ConfidenceState, TrajectoryState,
    ConsensusState, TemporalState, FluxPacket, Trend, Phase,
)


def random_zeitgeist() -> Zeitgeist:
    bloom = bytes(random.randint(0, 255) for _ in range(32))
    crdt = {random.randint(0, 10**18): random.randint(0, 10**18) for _ in range(random.randint(0, 5))}
    return Zeitgeist(
        precision=PrecisionState(
            deadband=random.uniform(0.001, 99999),
            funnel_pos=random.uniform(0, 1),
            snap_imminent=random.choice([True, False]),
        ),
        confidence=ConfidenceState(
            bloom=bloom,
            parity=random.randint(0, 255),
            certainty=random.uniform(0, 1),
        ),
        trajectory=TrajectoryState(
            hurst=random.uniform(0, 1),
            trend=random.choice([Trend.STABLE, Trend.RISING, Trend.FALLING, Trend.CHAOTIC]),
            velocity=random.uniform(-10, 10),
        ),
        consensus=ConsensusState(
            holonomy=random.uniform(0, 1),
            peer_agreement=random.uniform(0, 1),
            crdt_version=crdt,
        ),
        temporal=TemporalState(
            beat_pos=random.uniform(0, 1),
            phase=random.choice([Phase.IDLE, Phase.APPROACHING, Phase.SNAP, Phase.HOLD]),
            rhythm_coherence=random.uniform(0, 1),
        ),
    )


class TestMergeLaws(unittest.TestCase):
    """Prove CRDT semilattice laws with 100 random iterations each."""

    def test_commutativity(self):
        for _ in range(100):
            a = random_zeitgeist()
            b = random_zeitgeist()
            self.assertEqual(a.merge(b).to_dict(), b.merge(a).to_dict())

    def test_associativity(self):
        for _ in range(100):
            a = random_zeitgeist()
            b = random_zeitgeist()
            c = random_zeitgeist()
            self.assertEqual(
                a.merge(b).merge(c).to_dict(),
                a.merge(b.merge(c)).to_dict(),
            )

    def test_idempotency(self):
        for _ in range(100):
            a = random_zeitgeist()
            self.assertEqual(a.merge(a).to_dict(), a.to_dict())

    def test_json_roundtrip(self):
        for _ in range(50):
            zg = random_zeitgeist()
            encoded = zg.encode_json()
            decoded = Zeitgeist.decode_json(encoded)
            self.assertEqual(decoded.to_dict(), zg.to_dict())

    def test_alignment_valid(self):
        zg = random_zeitgeist()
        report = zg.check_alignment()
        self.assertTrue(report.aligned, f"Should be aligned: {report.violations}")

    def test_alignment_detects_violation(self):
        zg = random_zeitgeist()
        zg.precision.deadband = -1.0
        zg.confidence.certainty = 2.0
        zg.trajectory.hurst = 1.5
        zg.temporal.beat_pos = -0.5
        report = zg.check_alignment()
        self.assertFalse(report.aligned)
        self.assertGreaterEqual(len(report.violations), 4)

    def test_packet_roundtrip(self):
        zg = random_zeitgeist()
        pkt = FluxPacket(42, 99, b"hello flux", zg)
        encoded = pkt.encode()
        decoded = FluxPacket.decode(encoded)
        self.assertEqual(decoded.source, 42)
        self.assertEqual(decoded.target, 99)
        self.assertEqual(decoded.payload, b"hello flux")
        self.assertEqual(decoded.zeitgeist.to_dict(), zg.to_dict())


if __name__ == "__main__":
    unittest.main()
